src/main/java/graphql/schema/DataFetcherFactories.java

```diff
@@ -19,4 +22,26 @@
     public static <T> DataFetcherFactory<T> useDataFetcher(DataFetcher<T> dataFetcher) {
         return fieldDefinition -> dataFetcher;
     }
+
+    /**
+     * This helper function allows you to wrap an existing data fetcher and map the value once it completes.  It helps you handle
+     * values that might be {@link  java.util.concurrent.CompletionStage} returned values as well as plain old objects.
+     *
+     * @param delegateDataFetcher the original data fetcher that is present on a {@link graphql.schema.GraphQLFieldDefinition} say
+     * @param mapFunction         the bi function to apply to the original value
+     *
+     * @return a new data fetcher that wraps the provided data fetcher
+     */
+    public static DataFetcher wrapDataFetcher(DataFetcher delegateDataFetcher, BiFunction<DataFetchingEnvironment, Object, Object> mapFunction) {
+        return environment -> {
+            Object value = delegateDataFetcher.get(environment);
```

Comment:
Not sure this is really a concern here, but we also support DataFetcherResult and Optional values, both are automatically unwrapped when executed. We could do the same here ...

Reply:
I dont think there is any need.  We only want to know if its a CF so we can do the mapping function later.  If its Optional / DataFetcherResult then it will be handled by the execution strategy code later.

This is very much about composing a DF<T> to it can call back a mapping function with the original DF environment and the returned value


                def newFetcher = wrapDataFetcher(field.getDataFetcher(), { dfEnv, value ->
                    def directiveName = directiveEnv.directive.name
                    if (directiveName == "uppercase") {
                        return String.valueOf(value).toUpperCase()
                    } else if (directiveName == "lowercase") {
                        return String.valueOf(value).toLowerCase()
                    } else if (directiveName == "mixedcase") {
                        return toMixedCase(String.valueOf(value))
                    } else if (directiveName == "reverse") {
                        return String.valueOf(value).reverse()
                    }
                })